src/site/site.xml

```diff
@@ -46,6 +46,7 @@
       <!-- Will be added to every generated page -->
       <![CDATA[
         <script type="text/javascript" src="$relativePath/js/checkstyle.js"></script>
+        <script type="text/javascript" src="$relativePath/js/redirect.js"></script>
```

Comment:
Adding this script in `404.xml` was giving error, I tried different ways but it was failing so I had to use `site.xml` to inject `redirect.js` in `404.xml`

Reply:
redirect.js will be injected in all generated pages, so maybe we don't need 404.xml then. But I'm not sure about this, waiting for maintainers opinion.

Reply:
Js is in all existing pages. But we hit non existing. Such page use to exist.
So we need to recreate it as xml file with empty content. js in this page will trigger redirection.

In other words, page becomes real, but after page is loaded , js do redirection based on url content.

Reply:
Here is old website https://checkstyle.sourceforge.io/version/10.10.0/
So you can walk around to see a structure.

You can build site on local by
```
git checkout checkstyle-10.10.0
mvn clean site -Pno-validations 
```

We need to restore config_xxxx.xml files

Reply:
https://github.com/checkstyle/checkstyle/releases/tag/checkstyle-10.10.0

Such files can be found at:
https://github.com/checkstyle/checkstyle/tree/226ff1c57eaaca0bf6fcbf0674aece140c31119b/src/xdocs

Reply:
> We need to restore config_xxxx.xml files

From what I understood, are you trying to say if user comes to let's say `config_coding.html`, he should be redirected to https://checkstyle.org/checks/coding/ ?

Similarly if user comes to `config_coding#VariableDeclarationUsageDistance`, he should be redirected to https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

Im assuming, you're suggesting to restore all `config_xxx.xml` files and add redirection in them to redirect to their corresponding real page. 

But instead of restoring those files, we can add redirection in `404.xml` itself. When user will enter any `config_xxxx` url like https://checkstyle.org/config_coding.html#VariableDeclarationUsageDistance then he will hit the `404.xml` file and it will check the URL and redirect the user to its corresponding page, in this case: https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

We can add js in `404.xml` to check the URL and redirect the user, this way we will not need to restore `config_xxxx.xml` files. 

Correct me if I'm going in wrong direction.

Reply:
>config_coding.html, he should be redirected to https://checkstyle.org/checks/coding/ ?

Yes 

>config_coding#VariableDeclarationUsageDistance, he should be redirected to https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

Yes.

>But instead of restoring those files, we can add redirection in 404.xml itself

I doubt it will work as url will be ..../404.html , you will not have original url. But if you find a way to do it, I am ok with this.

Reply:
@rnveach @romani I created a new [personal repo](https://github.com/Zopsss/checkstyle-test) in which I kept all the CS docs. I also added the [404.html](https://github.com/Zopsss/checkstyle-test/blob/master/404.html) for the redirection.

Here's the site link: https://zopsss.github.io/checkstyle-test/

Currently I have added redirection only for some specific URLs which you can check from 404.html, you can test the website with any of the following URLs and it'll redirect you to CS' URLs.

https://zopsss.github.io/checkstyle-test/checks/

https://zopsss.github.io/checkstyle-test/config_coding.html

https://zopsss.github.io/checkstyle-test/config_coding.html#VariableDeclarationUsageDistance

If we inject similar JS in our 404.xml then we'll be able to redirect users to the original URLs. The only downside of this approach I can see is, if the body of the 404.html is empty then it'll not show any kind of message ( like page not found ) to the user.

For example:

![Screenshot 2024-11-28 121630](https://github.com/user-attachments/assets/55f5cece-8f9e-4437-bb51-b779ad2a3a40)

you can see in the screenshot that the URL is https://zopsss.github.io/checkstyle-test/something so it should give `404 page not found` error message, but the body of 404.html is empty so it does not display any kind of message.

We can fix this by designing our own error message, I would be happy to do that :)

Reply:
**UPDATE**

I have modified the 404.html and now it contains an error message, we can add something similar to that in our 404.xml file.

Reply:
I doubt that we can test the `404.xml` by generating website with `mvn clean site -Pno-validations` because then the URL will be different and the redirection won't happen, the only way to test is by releasing a new version.

Reply:
https://docs.github.com/en/pages/getting-started-with-github-pages/creating-a-custom-404-page-for-your-github-pages-site

Reply:
Yes, this documentation explains how we can use 404.html to show custom `404 page not found` error, just the way I used it in above comment.

Reply:
Push new code to let me see all details

Reply:
I have pushed the new code, I tried testing this code locally by changing the URLs to `localhost...` but it wasn't working as expected. To make it work, we will need to hit the URL from GH Pages deployment. So to check if it works, we will need to make a new release.

Is there any other way to test it?

Reply:
In the new code, I tried to add the JS by using `<script>` tag but I just noticed after generating the website, the `<script>` I added is removed:

![Screenshot 2024-11-30 095135](https://github.com/user-attachments/assets/044fc515-ed6a-4b5d-97e0-d24576188bac)

To inject JS in `404.html`, we can use this approach: https://github.com/checkstyle/checkstyle/commit/a7f0b2b8f80b1861467b223f1f0d5a11a516f33f

add the script in `site.xml` which will inject it in all the pages. But this will downgrade the performance of our website, we dont want this JS to be injected in all HTML pages, we want to inject it only in `404.html` page. Injecting this JS in all pages means that we will execute it every time a page loads, which might affect the loading time of our website.

Reply:
Micro code is ok

Reply:
I have pushed the micro code, but to test it, we will need to make a new release.